### PR TITLE
Feat/condo/doma-7447/payment model upgrade

### PR DESCRIPTION
**Pull request by AleX83Xpert**

### CODE REVIEW

File: apps/condo/domains/acquiring/schema/Payment.js [hunk 1]
```diff
@@ -139,9 +153,38 @@ const Payment = new GQLListSchema('Payment', {
...
+                    if (resolvedData[fieldPath] && !resolvedData['frozenInvoice']) {
+                        throw new GQLError(ERRORS.PAYMENT_NO_PAIRED_FROZEN_INVOICE, context)
+                    }
+                },
+            },
+        },
+
+        frozenInvoice: {
+            schemaDoc: 'Frozen invoice, used to resolving conflicts',
+            type: 'Json',
```
Comment (sitozzz) [inline apps/condo/domains/acquiring/schema/Payment.js:173]: Could it be strongly typed?

Comment (AleX83Xpert) [reply apps/condo/domains/acquiring/schema/Payment.js:173]: For now, it's not needed for this field. All frozen fields will be refactored.

**Review by VKislov — APPROVED** [review]

File: apps/condo/domains/acquiring/schema/Payment.js [hunk 2]
```diff
@@ -92,27 +106,27 @@ const Payment = new GQLListSchema('Payment', {
 
         advancedAt: {
             schemaDoc: 'Time at which money was advanced to recipient\'s account',
-            type: DateTimeUtc,
+            type: 'DateTimeUtc',
```
Comment (dkoviazin) [inline apps/condo/domains/acquiring/schema/Payment.js:109]: Why you change a type from wrapper to a keystone one?

File: apps/condo/domains/acquiring/schema/Payment.js [hunk 3]
```diff
@@ -92,27 +106,27 @@ const Payment = new GQLListSchema('Payment', {
...
             isRequired: false,
         },
 
         accountNumber: {
             schemaDoc: 'Payer\'s account number',
-            type: Text,
+            type: 'Text',
```
Comment (dkoviazin) [inline apps/condo/domains/acquiring/schema/Payment.js:115]: We can add a trim to wrapper

File: apps/condo/domains/marketplace/schema/Invoice.js [hunk 4]
```diff
@@ -133,16 +139,28 @@ const Invoice = new GQLListSchema('Invoice', {
...
             if (!((!!nextData.property && !!nextData.unitType && !!nextData.unitName) || !!nextData.accountNumber)) {
                 throw new GQLError(ERRORS.NO_RECEIVERS, context)
             }
 
             if (operation === 'update' && existingItem.status === INVOICE_STATUS_PAID) {
-                // TODO (DOMA-7447) prevent updating only for paid online (with Payment model) +TESTS
+                // Prevent updating only for paid online invoices (with Payment model)
                 // In the case of cash it's may be needed to uncheck the "paid" status
-                throw new GQLError(ERRORS.ALREADY_PAID, context)
+                const payments = await Payment.getAll(
```
Comment (dkoviazin) [inline apps/condo/domains/marketplace/schema/Invoice.js:151]: why getAll if you use id query?

File: apps/condo/migrations/20231023101614-0330_payment_frozeninvoice_payment_invoice_and_more.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+exports.up = async (knex) => {
```
Comment (dkoviazin) [inline apps/condo/migrations/20231023101614-0330_payment_frozeninvoice_payment_invoice_and_more.js:4]: you need to extend time limit for query, as we have lots of payments

File: apps/condo/domains/acquiring/schema/Payment.js (hunk 2 above)
Comment (AleX83Xpert) [reply apps/condo/domains/acquiring/schema/Payment.js:109]: All new models use string values of types and mapping during schema registration. As I remember, it's needed to help us to migrate to ks6 in the future. In the far future. In the very far future :)

File: apps/condo/domains/acquiring/schema/Payment.js (hunk 3 above)
Comment (AleX83Xpert) [reply apps/condo/domains/acquiring/schema/Payment.js:115]: If we will need some processing, we will create a custom field type.

File: apps/condo/domains/marketplace/schema/Invoice.js (hunk 4 above)
Comment (AleX83Xpert) [reply apps/condo/domains/marketplace/schema/Invoice.js:151]: There is no constraint "1 payment - 1 invoice". I am searching for payment by `invoice.id`. I'd like to find any payments for the given invoice.

File: apps/condo/migrations/20231023101614-0330_payment_frozeninvoice_payment_invoice_and_more.js (hunk 5 above)
Comment (AleX83Xpert) [reply apps/condo/migrations/20231023101614-0330_payment_frozeninvoice_payment_invoice_and_more.js:4]: Fixed

**Review by sitozzz — APPROVED** [review]

File: apps/condo/domains/acquiring/schema/Payment.js (hunk 1 above)
Comment (Alllex202) [inline apps/condo/domains/acquiring/schema/Payment.js:173]: In what format should values of this JSON field be? Any example?

File: apps/condo/domains/marketplace/schema/Invoice.js [hunk 6]
```diff
@@ -133,16 +139,28 @@ const Invoice = new GQLListSchema('Invoice', {
...
+                        invoice: { id: existingItem.id },
+                        status_in: [
+                            PAYMENT_PROCESSING_STATUS,
+                            PAYMENT_DONE_STATUS,
+                            PAYMENT_WITHDRAWN_STATUS,
+                        ],
+                    })
+                if (payments.length > 0) {
+                    throw new GQLError(ERRORS.ALREADY_PAID, context)
+                }
```
Comment (Alllex202) [inline apps/condo/domains/marketplace/schema/Invoice.js:163]: Mb it’s better to use `Payment.count` here so as not to get all the objects?

File: apps/condo/domains/marketplace/schema/Invoice.js [hunk 7]
```diff
@@ -133,16 +139,28 @@ const Invoice = new GQLListSchema('Invoice', {
...
+                const payments = await Payment.getAll(
+                    context,
+                    {
+                        invoice: { id: existingItem.id },
+                        status_in: [
+                            PAYMENT_PROCESSING_STATUS,
+                            PAYMENT_DONE_STATUS,
+                            PAYMENT_WITHDRAWN_STATUS,
+                        ],
+                    })
```
Comment (Alllex202) [inline apps/condo/domains/marketplace/schema/Invoice.js:160]: Do you need to filter deleted objects?

File: apps/condo/domains/acquiring/schema/Payment.js (hunk 1 above)
Comment (AleX83Xpert) [reply apps/condo/domains/acquiring/schema/Payment.js:173]: It's like the `frozenReceipt` field. All these frozen fields will be refactored in the future.

File: apps/condo/domains/marketplace/schema/Invoice.js (hunk 6 above)
Comment (AleX83Xpert) [reply apps/condo/domains/marketplace/schema/Invoice.js:163]: You right! Fixed.

**Review by dkoviazin — APPROVED** [review]

**Pull request merged**